# Variáveis e algoritmos

Renato Naville Watanabe

## Algoritmo


Sequência lógica de passos executados até atingir um objetivo ou solução de um problema.

Exemplos de situações onde uma sequência lógica (algoritmo) de passos é necessária:

- Fazer um bolo

- Construir um robô para explorar um local desconhecido

- Trocar uma lâmpada

Para a construção de um programa é necessário um conjunto de instruções colocadas em ordem sequencial lógica: um **Algoritmo**

## Ferramentas de computação científica

Existem diversas ferramentas para utilizadas em calculos matemáticos avançados.
- Matlab
- Mathematica
- Octave
- Scilab
- R
- [**Python**](https://nbviewer.jupyter.org/github/BMClab/bmc/blob/master/notebooks/PT-BR/PythonForScientificComputing_PT-BR.ipynb)
- Fortran
- Julia

Geralmente contam com bibliotecas de funções matemáticas prontas e recursos avançados.

## Spyder 
Ambiente de programação Python que permite a criação de novas funções/bibliotecas do usuário.

<img src="imagens/spyder.png" width="500">

O Spyder pode ser obtido com o pacote [Anaconda](https://www.anaconda.com/distribution/).

## Spyder

A interação do usuário com o Spyder pode ocorrer de duas formas distintas:

- Na **primeira forma**, os comando são digitados diretamente no *console* do Spyder:
    + Ao ser pressionada a tecla Enter, os comandos digitados são interpretados e imediatamente executados.
    + O Spyder (Python) funciona como uma  calculadora sofisticada.

- Na **segunda forma**, um conjunto de comandos é digitado em um arquivo texto:
    + Este arquivo  (script) pode ser  executado.
    + Neste modo, o Spyder funciona como um ambiente de programação.

## Spyder

Digitando o seguinte comando no console:

In [1]:
x = 2

estaremos criando uma variável real chamada $x$ cujo valor é igual a 2.

## Variável

Uma variável é um endereço da memória de acesso randômico (RAM), representada por um nome 
(rótulo), criado pelo usuário, cujo conteúdo pode se alterar no decorrer do programa.

Uma variável é composta por dois elementos:
    
  - **Identificador**: nome dado pelo programador à variável
  - **Conteúdo**: valor da variável
<img src="imagens/memvar2.png" width="400">

No caso anterior, x é o identificador (nome da variável) e 2 é o conteúdo (valor).

## Variável

Cada vez que mencionarmos o nome de uma variável, estaremos na verdade utilizando o seu conteúdo

In [2]:
y = x + 5
y

7

Esta operação define $y$ como sendo uma variável com valor igual ao valor de $x$ **mais** (+) cinco. Ou seja, $y$ tem um valor igual a 7

In [3]:
z = x*y
z

14

Neste caso, $z$ é igual à **multiplicação** (\*) dos valores guardados em $x$ e $y$, ou seja, $z$ é igual a 14.

In [4]:
w = z/x
w

7.0

Aqui, $w$ é igual à **divisão** (/) dos valores guardados em $z$ e $x$, ou seja, $w$
é igual a 7.

In [5]:
k = w-x
k

5.0

Aqui, $k$ é igual à **subtração** (-) do valor guardado em $x$ de $w$. Ou seja, $k$
é igual a 5.

In [6]:
m = k**3
m

125.0

Aqui, $m$ é igual ao valor guardado em $k$ **elevado** (\**) ao cubo. Ou seja, $m$ é igual a 125.

In [14]:
n = 10%3
n

1

Neste caso, $n$ tem o valor do **resto** (\%) da divisão entre 10 e 3.

## Operações aritméticas básicas

Portanto, as operações aritméticas básicas são:

- **multiplicação**: \*
- **divisão**: /
- **adição**: +
- **subtração**: -
- **potência**: \**
- **resto**: \%

## Ordem de operações

Assim como na matemática, em linguagens de programação a multiplicação e a divisão tem precedência sobre soma e subtração.

Assim, na operação:

$$\underbrace{\underbrace{4+\underbrace{3*4}_{12}}_{16}-2}_{14}$$

primeiramente é executada a multiplicação e depois a soma e a subtração na ordem em que elas aparecem.

In [22]:
4+3*4-2

14

A ordem de precedência das operações aritméticas básicas é:

- \** : potência
- \*, /, \% : multiplicação, divisão e resto
- +, - : adição, subtração

No caso de operações com a mesma precedência, prevalece a ordem na qual as operações foram escritas.

Uma maneira de mudar a ordem de execução das operações é utilizando parênteses. As operações dentro de parênteses tem precedência:

In [27]:
(4+3)*4-2

26

## Função

- O estudo de funções decorre da necessidade de:
    + Analisar fenômenos, visualizando o comportamento de um sistema.
    + Interpretar interdependências, entendendo como uma variável comporta-se com relação à outra.
    + Encontrar soluções de problemas.
    + Descrever regularidades.
    + Generalizar.

## Definição de uma função

- Uma função é uma regra segundo a qual, para cada elemento $x$ pertencente a um conjunto $A$ corresponde um único elemento $y$ pertencente a um conjunto $B$.

$$y = f(x)$$

- O conjunto $A$ é chamado domínio da função e o conjunto $B$ é o contra-domínio, ou imagem.




## Representação de uma função

- Uma função pode ser representada das seguintes formas: 
    + **Verbalmente** (descrevendo-a com palavras)
    + **Numericamente** (utilizando tabela de valores)
    + **Visualmente** (utilizando gráficos)
    + **Algebricamente** (utilizando uma fórmula explícita)

## Funções matemáticas

Para usar funções mais complexas, é necessário utilizar a biblioteca Numpy. Para poder utilizar esta biblioteca, é necessário importá-la:

In [7]:
import numpy as np

Esta biblioteca tem diversas funções que são facilmente utilizadas colocando 'np.' antes da função desejada. Por exemplo, a função logaritmo na base 10 é chamada com: 

In [8]:
np.log10(16)

1.2041199826559248

Se você quiser guardar o resultado em uma variável também é possível:

In [9]:
a = np.log10(16)
a

1.2041199826559248

E se quiser calcular o valor da função para uma variável é simples:

In [10]:
b = np.log10(k)
b

0.69897000433601886

## Funções matemáticas

Alguns exemplos de funções são:


**Exponencial**: exp($x$)=$e^x$ em que $e=2,7182...$ 

In [11]:
np.exp(-2)

0.1353352832366127

**Raiz quadrada**

In [12]:
s = np.sqrt(25)
s

5.0

**Funções trigonométricas**

In [13]:
d = np.sin(np.pi/2)
g = np.cos(np.pi/4)
print(d)
print(g)

1.0
0.707106781187


Nos dois últimos casos, para representar o valor de $\pi$, foi utilizado 'np.pi'.

## Referências

- Chalco, JM, *Slides de Bases Computacionais da Ciência*, (2014)
- Leite, S, *Slides de Bases Computacionais da Ciência*, (2018)
- [Marietto, MGB et al.; **Bases computacionais da Ciência** (2013)](http://prograd.ufabc.edu.br/images/pdf/bases_computacionais_livro.pdf).
- [Duarte, M, *Python para computação científica* (2017)](https://nbviewer.jupyter.org/github/BMClab/bmc/blob/master/notebooks/PT-BR/PythonForScientificComputing_PT-BR.ipynb)